ink/crates/phat_rollup_anchor_ink/src/traits/meta_transaction.rs

```diff
@@ -0,0 +1,173 @@
+    NonceTooLow,
+    IncorrectSignature,
+    PublicKeyNotMatch,
+    PublicKeyNotRegistered,
+    AccessControlError(AccessControlError),
+    RollupAnchorError(RollupAnchorError),
+}
+
+/// convertor from AccessControlError to MetaTxError
+impl From<AccessControlError> for MetaTransactionError {
+    fn from(error: AccessControlError) -> Self {
+        MetaTransactionError::AccessControlError(error)
+    }
+}
+
+/// convertor from RollupAnchorError to MetaTxError
+impl From<RollupAnchorError> for MetaTransactionError {
+    fn from(error: RollupAnchorError) -> Self {
+        MetaTransactionError::RollupAnchorError(error)
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, scale::Encode, scale::Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub struct ForwardRequest {
+    pub from: AccountId,
+    pub nonce: Nonce,
+    pub data: Vec<u8>,
+}
+
+#[derive(Default, Debug)]
+#[openbrush::storage_item]
+pub struct Data {
+    nonces: Mapping<AccountId, Nonce>,
+    ecdsa_public_keys: Mapping<AccountId, Vec<u8>>,
+}
+
+#[openbrush::trait_definition]
+pub trait MetaTransaction:
+    Storage<Data> + EventBroadcaster + access_control::Internal + RollupAnchor
+{
+    #[ink(message)]
+    fn get_ecdsa_public_key(&self, from: AccountId) -> [u8; 33] {
+        self.data::<Data>()
+            .ecdsa_public_keys
+            .get(&from)
+            .map(|k| k.try_into().unwrap_or([0; 33]))
+            .unwrap_or([0; 33])
+    }
+
+    #[ink(message)]
+    #[openbrush::modifiers(access_control::only_role(MANAGER_ROLE))]
+    fn register_ecdsa_public_key(
+        &mut self,
+        from: AccountId,
+        ecdsa_public_key: [u8; 33],
+    ) -> Result<(), MetaTransactionError> {
+        self.data::<Data>()
+            .ecdsa_public_keys
+            .insert(&from, &ecdsa_public_key.into());
+        Ok(())
+    }
+
+    #[ink(message)]
+    fn prepare(
+        &self,
+        from: AccountId,
+        data: Vec<u8>,
+    ) -> Result<(ForwardRequest, Hash), MetaTransactionError> {
+        let nonce = self.get_nonce(from);
+
+        let request = ForwardRequest { from, nonce, data };
+        let mut hash = <Blake2x256 as HashOutput>::Type::default();
+        ink::env::hash_encoded::<Blake2x256, _>(&request, &mut hash);
+
+        Ok((request, hash.into()))
+    }
+
+    fn get_nonce(&self, from: AccountId) -> Nonce {
+        self.data::<Data>().nonces.get(&from).unwrap_or(0)
+    }
+
+    fn verify(
+        &self,
+        request: &ForwardRequest,
+        signature: &[u8; 65],
+    ) -> Result<(), MetaTransactionError> {
+        let ecdsa_public_key: [u8; 33] = self
+            .data::<Data>()
+            .ecdsa_public_keys
+            .get(&request.from)
+            .ok_or(MetaTransactionError::PublicKeyNotRegistered)?
+            .try_into()
+            .map_err(|_| MetaTransactionError::PublicKeyNotRegistered)?;
+
+        let nonce_from = self.get_nonce(request.from);
+
+        if request.nonce < nonce_from {
```

Comment:
`assert(request.nonce == nonce_from)`

Reply:
My mentor said me "a smart contract should not never panic", it's why return an Error seems better to me :)
I test "request.nonce < nonce_from" instead of "request.nonce == nonce_from" because it was done like this in evm side.

Reply:
If you want I change the test  "request.nonce < nonce_from" by "request.nonce == nonce_from", I can do it but it will not be align anymore with teh evm version.

Reply:
I meant the nonce must match the next expected nonce. So if the give nonce is less, it's an error. On the other hand, if it's larger than expected, it should also give an error.

I didn't mean to put an assert here.

Reply:
I did the update. 
I recommend to update also the solidity contract to be aligned.